cgo/lapacke/generate_lapacke.go

```diff
@@ -0,0 +1,554 @@
+// Copyright ©2016 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
+
+// generate_lapacke creates a lapacke.go file from the provided C header file
+// with optionally added documentation from the documentation package.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/cznic/cc"
+
+	"github.com/gonum/internal/binding"
+)
+
+const (
+	header        = "lapacke.h"
+	documentation = "../native"
+	target        = "lapacke.go"
+
+	prefix = "LAPACKE_"
+	suffix = "_work"
+)
+
+const (
+	elideRepeat = true
+	noteOrigin  = false
+)
+
+var skip = map[string]bool{
+	// Deprecated.
+	"LAPACKE_cggsvp_work": true,
+	"LAPACKE_dggsvp_work": true,
+	"LAPACKE_sggsvp_work": true,
+	"LAPACKE_zggsvp_work": true,
+	"LAPACKE_cggsvd_work": true,
+	"LAPACKE_dggsvd_work": true,
+	"LAPACKE_sggsvd_work": true,
+	"LAPACKE_zggsvd_work": true,
+	"LAPACKE_cgeqpf_work": true,
+	"LAPACKE_dgeqpf_work": true,
+	"LAPACKE_sgeqpf_work": true,
+	"LAPACKE_zgeqpf_work": true,
+}
+
+// needsInt is a list of routines that need to return the integer info value and
+// and cannot convert to a success boolean.
+var needsInt = map[string]bool{
+	"hseqr": true,
+	"geev":  true,
+	"geevx": true,
+}
+
+// allUplo is a list of routines that allow 'A' for their uplo argument.
+// The list keys are truncated by one character to cover all four numeric types.
+var allUplo = map[string]bool{
+	"lacpy": true,
+}
+
+var cToGoType = map[string]string{
+	"char":           "byte",
+	"int_must":       "int",
+	"int_must32":     "int32",
+	"int":            "bool",
+	"float":          "float32",
+	"double":         "float64",
+	"float complex":  "complex64",
+	"double complex": "complex128",
+}
+
+var cToGoTypeConv = map[string]string{
+	"int_must":       "int",
+	"int":            "isZero",
+	"float":          "float32",
+	"double":         "float64",
+	"float complex":  "complex64",
+	"double complex": "complex128",
+}
+
+var cgoEnums = map[string]*template.Template{}
+
+var byteTypes = map[string]string{
+	"compq": "lapack.Comp",
```

Comment:
It does not belong into this PR but I'll forget otherwise: I may have mentioned it before but I've been thinking that we could avoid referencing `lapack` from `lapacke` by simply using `byte` for these parameters. This PR enables to fine-tune better the type of the parameters but often we have to do a type conversion anyway. Moreover, for example in case of `side` it is more complicated by the fact that the type depends also on the routine (for example, I use `side lapack.EVSide` in `native.Dgebak`, not `blas.Side`).